rialto_airflow/harvest/wos.py

```diff
@@ -77,28 +79,82 @@ def harvest(snapshot: Snapshot, limit=None) -> Path:
     return jsonl_file
 
 
+def fill_in(snapshot: Snapshot, jsonl_file: Path):
+    """Harvest WebOfScience data for DOIs from other publication sources."""
+    count = 0
+    with jsonl_file.open("a") as jsonl_output:
+        with get_session(snapshot.database_name).begin() as select_session:
+            stmt = (
+                select(Publication.doi)  # type: ignore
+                .where(Publication.doi.is_not(None))  # type: ignore
+                .where(Publication.wos_json.is_(None))
+                .execution_options(yield_per=50)
+            )
+
+            for rows in select_session.execute(stmt).partitions():
+                # since the query uses yield_per=50 we will be looking up 50 DOIs at a time
+                dois = [row["doi"] for row in rows]
+
+                logging.info(f"looking up DOIs {dois}")
+                for wos_pub in publications_from_dois(dois):
+                    doi = normalize_doi(wos_pub.get("doi"))
+                    if doi is None:
+                        logging.warning("unable to determine what DOI to update")
+                        continue
+
+                    with get_session(snapshot.database_name).begin() as update_session:
+                        update_stmt = (
+                            update(Publication)  # type: ignore
+                            .where(Publication.doi == doi)
+                            .values(wos_json=wos_pub)
+                        )
+                        update_session.execute(update_stmt)
+
+                    count += 1
+                    jsonl_output.write(json.dumps(wos_pub) + "\n")
+
+    logging.info(f"filled in {count} publications")
+
+    return snapshot.path
+
+
 def orcid_publications(orcid) -> Generator[dict, None, None]:
     """
     A generator that returns publications associated with a given ORCID.
     """
-
-    # For API details see: https://api.clarivate.com/swagger-ui/
-
     # WoS doesn't recognize ORCID URIs which are stored in User table
     if m := re.match(r"^https?://orcid.org/(.+)$", orcid):
         orcid = m.group(1)
 
+    yield from _wos_api(f"AI={orcid}")
+
+
+def publications_from_dois(dois: list[str]) -> Generator[dict, None, None]:
+    """
+    A generator that returns publications associated a list of DOIs.
+    """
+    for doi_batch in batched(dois, n=50):
+        yield from _wos_api(f"DO=({' '.join(doi_batch)})")
+
+
```

Comment:
Since the paging logic for the WoS API was the same for looking up by DOI it was pulled out into `_wos_api()` where it could be reused.